File: python/bi2de.py
import numpy as np

# Binary vector to decimal scalar converter, good to 64-bits.
def bi2de(bin,flag = 'left-msb'):
    if bin.ndim == 2:
        d = np.empty(bin.shape[1])
        for i in xrange(0, bin.shape[1]):
            d[i] = bi2de1(bin[:,i],flag)
    else:
        d = bi2de1(bin,flag)
    return d


def bi2de1(bin,flag = 'left-msb'):
    if flag.lower() == 'right-msb':
        bin = bin[ : : -1]
    return bin.dot(1 << np.arange(bin.shape[-1] - 1, -1, -1,dtype=np.uint64))

File: matlab/bi2de.m
function d = bi2de(b, p, f)
% Usage: b = bi2de(b, p, f)
%
% Convert binary vectors to decimal numbers.
%
% b..........binary vector to convert
% p..........optional base of vector to convert from (default=2)
% f..........flag defining the target orientation
%            one of : 'left-msb', 'right-msb'
%            default is 'right'msb'
%

  switch (nargin)
    case 1
      p = 2;
      f = "right-msb";
    case 2
      if (ischar (p))
        f = p;
        p = 2;
      else
        f = "right-msb";
      end
    case 3
      if (ischar (p))
        tmp = f;
        f = p;
        p = tmp;
      end
    otherwise
      help bi2de
      return
  end

  if (~ (all (b(:) == fix (b(:))) && all (b(:) >= 0) && all (b(:) < p)))
    error ("bi2de: all elements of B must be integers in the range [0,P-1]");
  end

  if (strcmp (f, "left-msb"))
    b = b(:,size (b, 2):-1:1);
  elseif (~strcmp (f, "right-msb"))
    error ("bi2de: invalid option '%s'", f);
  end

  if (length (b) == 0)
    d = [];
  else
    d = b * (p .^ [0:(columns (b) - 1)]');
  end

function y = columns(x)
  y = size(x,2);
